packages/go_router/lib/src/router.dart

```diff
@@ -192,6 +194,32 @@ class GoRouter extends ChangeNotifier implements RouterConfig<RouteMatchList> {
   /// Navigate to a URI location w/ optional query parameters, e.g.
   /// `/family/f2/person/p1?color=blue`
   void go(String location, {Object? extra}) {
+    final FutureOr<bool> Function(BuildContext context)? onExit =
+        routerDelegate.currentConfiguration.isNotEmpty
+            ? routerDelegate.currentConfiguration.last.route.onExit
```

Comment:
why do we only call on exit on the last one?

Reply:
You're right, I was thinking too much about pop.
So I should compute the new match list, find where it starts to differ and then call each onExit (if any) starting from the end?

Reply:
yes I think so.

Reply:
I've implemented this logic and updated the test to make sure onExit is called for all routes in [Call onExit for every exiting route when using go](https://github.com/flutter/packages/pull/3709/commits/ade1f0875f38ab9fcd9f8c3d1a5f04f48429bb5e).

For `replace` and `pushReplacement`, currently the `onExit` will not be called. In the documentation for `replace`, it says:

> Replaces the top-most page of the page stack with the given one but treats it as the same page.
The page key will be reused. This will preserve the state and not run any page animation.

So would it be fair to say that `onExit` should only be called when you use `pushReplacement`? Or do you want me to add it in both cases?

Reply:
Hi @chunhtai , I went ahead and implemented the call to `onExit` when using `pushReplacement` but not in `replace`.
[Eval onExit when using pushReplacement](https://github.com/flutter/packages/pull/3709/commits/0eacb09ac64332fdca6d09b61994dfbab61095a7)